core/sdk-core/src/main/java/software/amazon/awssdk/core/internal/http/pipeline/stages/MakeAsyncHttpRequestStage.java

```diff
@@ -74,16 +74,61 @@ public MakeAsyncHttpRequestStage(TransformingAsyncResponseHandler<Response<Outpu
                                      HttpClientDependencies dependencies) {
         this.responseHandler = responseHandler;
         this.futureCompletionExecutor =
-            dependencies.clientConfiguration().option(SdkAdvancedAsyncClientOption.FUTURE_COMPLETION_EXECUTOR);
+                dependencies.clientConfiguration().option(SdkAdvancedAsyncClientOption.FUTURE_COMPLETION_EXECUTOR);
         this.sdkAsyncHttpClient = dependencies.clientConfiguration().option(SdkClientOption.ASYNC_HTTP_CLIENT);
         this.apiCallAttemptTimeout = dependencies.clientConfiguration().option(SdkClientOption.API_CALL_ATTEMPT_TIMEOUT);
         this.timeoutExecutor = dependencies.clientConfiguration().option(SdkClientOption.SCHEDULED_EXECUTOR_SERVICE);
     }
 
     @Override
-    public CompletableFuture<Response<OutputT>> execute(SdkHttpFullRequest request,
-                                                        RequestExecutionContext context) throws Exception {
-        return executeHttpRequest(request, context);
+    public CompletableFuture<Response<OutputT>> execute(CompletableFuture<SdkHttpFullRequest> requestFuture,
+                                                        RequestExecutionContext context) {
+        CompletableFuture<Response<OutputT>> toReturn = new CompletableFuture<>();
+
+        // Setup the cancellations. If the caller fails to provide a request, forward the exception to the future we
+        // return
+        requestFuture.whenComplete((r, t) -> {
+            if (t != null) {
+                toReturn.completeExceptionally(t);
+            }
+        });
```

Comment:
Why not CompletableFutureUtils.forwardExceptionTo here and below?

Reply:
Let me try it out. I was having a hard time to trying to reconcile the cancellation here and also ensuring that exceptions from the interceptor callbacks (e.g. `afterTransmission`) were propagated correctly back to the caller future and this is what I landed on.